### ISSUE TITLE
False Negative: `abstract`, `native`, wrong-signature, and one-sided override forms are not fully incorporated into the `equals`/`hashCode` check.

### ISSUE THREAD

**Issue by Carlson-JLQ** [google/error-prone#5631]
# False Negative: `abstract`, `native`, wrong-signature, and one-sided override forms are not fully incorporated into the `equals`/`hashCode` check.

Version
2.48.0

## Checker
- Checker id: `EqualsHashCode`
- Checker description: Detects classes that override equals(Object) in a non-trivial way but do not override hashCode().

## Description of the false negative
`diff/output` marks 15 false negatives for `EqualsHashCode`. Across these samples, the benchmark is still exercising the same underlying bug pattern, but the implementation appears to lose track of it once the source deviates from one canonical shape. The recurring themes are suppression or annotations, delegation, signature edge cases.

### `PosCase5.java`
Source: `output/checkstyle/EqualsHashCodeCheck/src/main/java/scensct/core/pos/PosCase5.java`

This class only declares one side of the `equals`/`hashCode` pair abstractly instead of providing a concrete implementation. The missed report suggests the implementation is treating abstract declarations as if they fully satisfied the rule's counterpart requirement. That is also how the benchmark classifies the sample: A class defines a concrete equals(Object) method and any zero-parameter hashCode method defined is abstract should be flagged as violation.

```java
// A class defines a concrete equals(Object) method and any zero-parameter hashCode method defined is abstract should be flagged as violation.
package scensct.core.pos;

public abstract class PosCase5 {
    private int value;

    // Concrete equals method defined
    public boolean equals(Object obj) { // [REPORTED LINE]
        if (obj instanceof PosCase5) {
            return this.value == ((PosCase5) obj).value;
        }
        return false;
    }

    // Abstract hashCode method (no body, not native)
    public abstract int hashCode();
}
```

### `PosCase6.java`
Source: `output/checkstyle/EqualsHashCodeCheck/src/main/java/scensct/core/pos/PosCase6.java`

This class only declares one side of the `equals`/`hashCode` pair abstractly instead of providing a concrete implementation. The missed report suggests the implementation is treating abstract declarations as if they fully satisfied the rule's counterpart requirement. That is also how the benchmark classifies the sample: A class defines a native equals(Object) method and any zero-parameter hashCode method defined is abstract should be flagged as violation.

```java
// A class defines a native equals(Object) method and any zero-parameter hashCode method defined is abstract should be flagged as violation.
package scensct.core.pos;

public abstract class PosCase6 {
    private int value;

    // Native equals method defined
    public native boolean equals(Object obj); // [REPORTED LINE]

    // Abstract hashCode method (no body, not native)
    public abstract int hashCode();
}
```

### `PosCase1.java`
Source: `output/codeql/InconsistentEqualsHashCode/src/main/java/scensct/core/pos/PosCase1.java`

The sample keeps the benchmark's target pattern but expresses it in a slightly non-canonical source form. The risky pattern is still present in the code, so the absence of a report points to a matching gap rather than a benign variant. That is also how the benchmark classifies the sample: A class that overrides hashCode but does not override equals should be flagged as a violation.

```java
// A class that overrides hashCode but does not override equals should be flagged as a violation.
package scensct.core.pos;

public class PosCase1 { // [REPORTED LINE]
    private int id;
    
    @Override
    public int hashCode() {
        return id;
    }
    // No equals method overridden, violating the contract.
}
```

### `PosCase3.java`
Source: `output/codeql/InconsistentEqualsHashCode/src/main/java/scensct/core/pos/PosCase3.java`

The sample keeps the benchmark's target pattern but expresses it in a slightly non-canonical source form. The risky pattern is still present in the code, so the absence of a report points to a matching gap rather than a benign variant. That is also how the benchmark classifies the sample: A class that does not override equals(Object) but declares at least one method named hashCode should be flagged as a violation.

```java
// A class that does not override equals(Object) but declares at least one method named hashCode should be flagged as a violation.
package scensct.core.pos;

public class PosCase3 { // [REPORTED LINE]
    private int id;
    
    // This method overrides Object.hashCode.
    @Override
    public int hashCode() {
        return id;
    }
    // No equals method overridden, violating the contract.
}
```

### `PosCase1.java`
Source: `output/pmd/OverrideBothEqualsAndHashcode/src/main/java/scensct/core/pos/PosCase1.java`

The only interesting behavior here is an `equals(Object)` override that immediately delegates to `super.equals(obj)`. The benchmark still expects a report because the class shape violates the surrounding contract even though the method body is trivial. That is also how the benchmark classifies the sample: A non-Comparable class with equals and hashCode should be flagged as overriding only one of them.

```java
// A non-Comparable class with equals and hashCode should be flagged as overriding only one of them.
package scensct.core.pos;

public class PosCase1 {
    // Non-Comparable class with valid equals and hashCode methods
    static class MyClass {
        @Override
        public boolean equals(Object obj) { // [REPORTED LINE]
            return super.equals(obj);
        }
        
        // hashCode removed to override only equals
    }
    
    public static void main(String[] args) {
        // Minimal instantiation to make class usable
        MyClass obj = new MyClass();
        System.out.println(obj.equals(null));
    }
}
```

### `PosCase1.java`
Source: `output/sonarqube/EqualsOverriddenWithHashCodeCheck/src/main/java/scensct/core/pos/PosCase1.java`

The only interesting behavior here is an `equals(Object)` override that immediately delegates to `super.equals(obj)`. The benchmark still expects a report because the class shape violates the surrounding contract even though the method body is trivial. That is also how the benchmark classifies the sample: A class defines a proper public non-static boolean equals(Object) method but does not define any hashCode method should be flagged as Positive.

```java
// A class defines a proper public non-static boolean equals(Object) method but does not define any hashCode method should be flagged as Positive.
package scensct.core.pos;

public class PosCase1 {
    @Override
    public boolean equals(Object obj) { // [REPORTED LINE]
        return super.equals(obj);
    }
}
```

### `PosCase10.java`
Source: `output/sonarqube/EqualsOverriddenWithHashCodeCheck/src/main/java/scensct/core/pos/PosCase10.java`

The class defines an `equals`-like overload, so the rule has to distinguish a real `equals(Object)` override from a same-name method with the wrong signature. The dangerous part is that there is no real `equals(Object)` implementation, only an overload with the wrong signature. That is also how the benchmark classifies the sample: A class defines a proper int hashCode() method and defines at least one public non-static boolean equals method with exactly one parameter, but the parameter type of every such method is not java.lang.Object should be flagged as Positive.

```java
// A class defines a proper int hashCode() method and defines at least one public non-static boolean equals method with exactly one parameter, but the parameter type of every such method is not java.lang.Object should be flagged as Positive.
package scensct.core.pos;

public class PosCase10 {
    @Override
    public int hashCode() { // [REPORTED LINE]
        return super.hashCode();
    }

    // Public non-static boolean equals with wrong parameter type
    public boolean equals(String obj) {
        return false;
    }
}
```

### `PosCase2.java`
Source: `output/sonarqube/EqualsOverriddenWithHashCodeCheck/src/main/java/scensct/core/pos/PosCase2.java`

The sample keeps the benchmark's target pattern but expresses it in a slightly non-canonical source form. The risky pattern is still present in the code, so the absence of a report points to a matching gap rather than a benign variant. That is also how the benchmark classifies the sample: A class defines a proper int hashCode() method but does not define any equals method should be flagged as Positive.

```java
// A class defines a proper int hashCode() method but does not define any equals method should be flagged as Positive.
package scensct.core.pos;

public class PosCase2 {
    @Override
    public int hashCode() { // [REPORTED LINE]
        return super.hashCode();
    }
}
```

### `PosCase3.java`
Source: `output/sonarqube/EqualsOverriddenWithHashCodeCheck/src/main/java/scensct/core/pos/PosCase3.java`

The only interesting behavior here is an `equals(Object)` override that immediately delegates to `super.equals(obj)`. The benchmark still expects a report because the class shape violates the surrounding contract even though the method body is trivial. That is also how the benchmark classifies the sample: A class defines a proper public non-static boolean equals(Object) method and defines at least one method named hashCode, but every such hashCode method has one or more parameters should be flagged as Positive.

```java
// A class defines a proper public non-static boolean equals(Object) method and defines at least one method named hashCode, but every such hashCode method has one or more parameters should be flagged as Positive.
package scensct.core.pos;

public class PosCase3 {
    @Override
    public boolean equals(Object obj) { // [REPORTED LINE]
        return super.equals(obj);
    }

    // hashCode with parameter, not a valid zero-parameter hashCode
    public int hashCode(String s) {
        return s.hashCode();
    }
}
```

### `PosCase4.java`
Source: `output/sonarqube/EqualsOverriddenWithHashCodeCheck/src/main/java/scensct/core/pos/PosCase4.java`

The only interesting behavior here is an `equals(Object)` override that immediately delegates to `super.equals(obj)`. The benchmark still expects a report because the class shape violates the surrounding contract even though the method body is trivial. That is also how the benchmark classifies the sample: A class defines a proper public non-static boolean equals(Object) method and defines at least one zero-parameter method named hashCode, but every such method has a return type other than int should be flagged as Positive.

```java
// A class defines a proper public non-static boolean equals(Object) method and defines at least one zero-parameter method named hashCode, but every such method has a return type other than int should be flagged as Positive.
package scensct.core.pos;

public class PosCase4 {
    @Override
    public boolean equals(Object obj) { // [REPORTED LINE]
        return super.equals(obj);
    }

    // Zero-parameter hashCode with non-int return type
    public long hashCode(String ignored) {
        return 0L;
    }
}
```

### `PosCase5.java`
Source: `output/sonarqube/EqualsOverriddenWithHashCodeCheck/src/main/java/scensct/core/pos/PosCase5.java`

The class defines an `equals`-like overload, so the rule has to distinguish a real `equals(Object)` override from a same-name method with the wrong signature. The dangerous part is that there is no real `equals(Object)` implementation, only an overload with the wrong signature. That is also how the benchmark classifies the sample: A class defines a proper int hashCode() method and defines at least one equals method, but every such equals method is non-public should be flagged as Positive.

```java
// A class defines a proper int hashCode() method and defines at least one equals method, but every such equals method is non-public should be flagged as Positive.
package scensct.core.pos;

public class PosCase5 {
    @Override
    public int hashCode() { // [REPORTED LINE]
        return super.hashCode();
    }

    // Non-public equals method
    private boolean equals(String obj) {
        return false;
    }
}
```

### `PosCase6.java`
Source: `output/sonarqube/EqualsOverriddenWithHashCodeCheck/src/main/java/scensct/core/pos/PosCase6.java`

The class defines an `equals`-like overload, so the rule has to distinguish a real `equals(Object)` override from a same-name method with the wrong signature. The dangerous part is that there is no real `equals(Object)` implementation, only an overload with the wrong signature. That is also how the benchmark classifies the sample: A class defines a proper int hashCode() method and defines at least one public equals method, but every such public equals method is static should be flagged as Positive.

```java
// A class defines a proper int hashCode() method and defines at least one public equals method, but every such public equals method is static should be flagged as Positive.
package scensct.core.pos;

public class PosCase6 {
    @Override
    public int hashCode() { // [REPORTED LINE]
        return super.hashCode();
    }

    // Static equals method, not a valid instance equals
    public static boolean equals(String obj) {
        return false;
    }
}
```

### `PosCase7.java`
Source: `output/sonarqube/EqualsOverriddenWithHashCodeCheck/src/main/java/scensct/core/pos/PosCase7.java`

The sample keeps the benchmark's target pattern but expresses it in a slightly non-canonical source form. The risky pattern is still present in the code, so the absence of a report points to a matching gap rather than a benign variant. That is also how the benchmark classifies the sample: A class defines a proper int hashCode() method and defines at least one public non-static method, but no such method has the simple name "equals" should be flagged as Positive.

```java
// A class defines a proper int hashCode() method and defines at least one public non-static method, but no such method has the simple name "equals" should be flagged as Positive.
package scensct.core.pos;

public class PosCase7 {
    @Override
    public int hashCode() { // [REPORTED LINE]
        return super.hashCode();
    }

    // Public non-static method with name other than "equals"
    public void someMethod() {}
}
```

### `PosCase8.java`
Source: `output/sonarqube/EqualsOverriddenWithHashCodeCheck/src/main/java/scensct/core/pos/PosCase8.java`

The sample keeps the benchmark's target pattern but expresses it in a slightly non-canonical source form. The risky pattern is still present in the code, so the absence of a report points to a matching gap rather than a benign variant. That is also how the benchmark classifies the sample: A class defines a proper int hashCode() method and defines at least one public non-static method named equals, but every such method has a return type other than boolean should be flagged as Positive.

```java
// A class defines a proper int hashCode() method and defines at least one public non-static method named equals, but every such method has a return type other than boolean should be flagged as Positive.
package scensct.core.pos;

public class PosCase8 {
    @Override
    public int hashCode() { // [REPORTED LINE]
        return super.hashCode();
    }

    // Public non-static equals with non-boolean return type
    public int equals(String obj) {
        return 0;
    }
}
```

### `PosCase9.java`
Source: `output/sonarqube/EqualsOverriddenWithHashCodeCheck/src/main/java/scensct/core/pos/PosCase9.java`

The sample keeps the benchmark's target pattern but expresses it in a slightly non-canonical source form. The risky pattern is still present in the code, so the absence of a report points to a matching gap rather than a benign variant. That is also how the benchmark classifies the sample: A class defines a proper int hashCode() method and defines at least one public non-static boolean equals method, but every such method does not have exactly one parameter should be flagged as Positive.

```java
// A class defines a proper int hashCode() method and defines at least one public non-static boolean equals method, but every such method does not have exactly one parameter should be flagged as Positive.
package scensct.core.pos;

public class PosCase9 {
    @Override
    public int hashCode() { // [REPORTED LINE]
        return super.hashCode();
    }

    // Public non-static boolean equals with zero parameters
    public boolean equals() {
        return false;
    }
}
```

## Cause analysis
The rule is documented as follows: Detects classes that override equals(Object) in a non-trivial way but do not override hashCode().

In this set the problem shows up around suppression or annotations, delegation, signature edge cases. The missed cases suggest that the implementation is narrower than the rule description and too dependent on one preferred AST shape.

Once the same behavior is expressed through a helper, an overload, a modifier such as `abstract` or `native`, a different comment position, or a slightly rearranged control-flow structure, the report disappears even though the benchmark still treats the case as in-scope.

For `EqualsHashCode`, the practical improvement would be to generalize the match so it follows the underlying behavior rather than one exact spelling of it.

## References
None known.
